File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.util.Scanner;

public class Main {
    private static int ANSWER = 0;
    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            int n = sc.nextInt();
            int s = sc.nextInt();
            int[] array = new int[n];

            for (int i = 0; i < n; i++) {
                array[i] = sc.nextInt();
            }

            dfs(array, 0, 0, s);

            System.out.println(s == 0 ? ANSWER - 1 : ANSWER);
        }
    }

    private static void dfs(int[] array, int index, int sum, int target) {
        if (index == array.length) {
            if (sum == target) {
                ANSWER++;
            }
            return;
        }

        dfs(array, index+1, sum + array[index], target);
        dfs(array, index+1, sum, target);
    }
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val split = readLine().split(" ")
    val n = split[0].toInt()
    val target = split[1].toInt()
    val numArray = readLine().split(" ").map { it.toInt() }.toIntArray()
    var answer = 0

    fun backtracking(index:Int, result: Int) {
        if (index == n) {
            if (result == target)
                answer++
            return
        }

        backtracking(index+1, result + numArray[index])
        backtracking(index+1, result)
    }

    backtracking(0, 0)
    // target == 0일 때, 1을 빼주는 이유는 공집합일 경우를 제외하기 위해서.
    println(if (target == 0) answer-1 else answer)
}